src/wallet/spend.cpp

```diff
@@ -105,6 +105,54 @@ void CoinsResult::clear()
     other.clear();
 }
 
+// Fetch the coin control selected inputs and wrap them inside a COutput.
+// Coins could be internal (from the wallet) or external.
+BResult<CoinsResult> FetchSelectedInputs(const CWallet& wallet, const CCoinControl& coin_control,
```

Comment:
In 23448cae6f4a5f00766df9220860b16311266417 "wallet: encapsulate pre-selected-inputs lookup into its own function"

I think this should return a `SelectionResult` rather than `CoinsResult`. It would elimiate the second iteration inside of `SelectCoins`.

It could also return the selected value too so the caller could reduce the target before `SelectCoins` and eliminate an additional iteration of the preselcted inputs.

Reply:
good thought 👍🏼 

Most likely, will need to do this change incrementally (right after 23448cae) because if `FetchSelectedInputs` returns a `SelectionResult`, will need to add a method for merging two `SelectionResult` objects.

We are currently creating an `OutputGroup` at the beginning of `SelectCoins` with the pre-set-inputs just to merge it at the end of the function with the wallet's selection algorithm result. The reason is that `SelectResult` only has a function `AddInputs(OutputGroups)` and not an `AddInputs(set<COutputs>)` nor a merge method.

Side note, based on a quick code review: the selection result built from the combination of pre-set-inputs and the wallet selected ones should be having an inaccurate waste calculation -- will do some tests in another PR --).